### PR TITLE
Normal Debits/Normal Credits - Closes #8

### LINKED ISSUE #8: Need a test to make sure journal entry will balance with valid input
This issue is concerned with the following code:

```
public record JournalEntryLine
{
        public JournalEntryLine(Account account, decimal amount)
        {
            Account = account;
            Amount = amount;
        }

        public Account Account { get; }
        public decimal Amount { get; }
}
```
and
```
public bool PostEntry(JournalEntry journalEntry)
{
    return journalEntry.Credits.Sum(credit => credit.Amount) == journalEntry.Debits.Sum(debit => debit.Amount);
}
```

Currently JournalEntryLine accepts signed decimal value for its Amount property. ``PostEntry`` will fail if Credit = 100 and Debit = -100.

### PR DESCRIPTION
**Pull request by maratmuginov**
Accounts have normal debits / normal credits to validate transactions in accordance to double entry accounting.

### CODE REVIEW

File: source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs [hunk 1]
```diff
@@ -39,5 +43,28 @@ public JournalEntry(DateTime date, [NotNull] List<JournalEntryLine> debits, [Not
...
+        private static void EnsureNormalBalances(ICollection<JournalEntryLine> debits, 
+            ICollection<JournalEntryLine> credits)
+        {
+            //TODO : Write helpful exception messages.
+            if (debits.Any(dr => dr.Account.NormalBalance is not NormalBalance.Debit || dr.Amount < 0))
+                throw new ArgumentException();
+
+            if (credits.Any(cr => cr.Account.NormalBalance is not NormalBalance.Credit || cr.Amount > 0))
+                throw new ArgumentException();
+        }
```
Comment (itiannerona) [inline source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs]: May I suggest this code instead?

```public interface IJournalEntryLine
{
    Account Account;
    decimal Amount;
}

public JournalDebitLine : IJournalEntryLine
{
    public JournalDebitLine(Account account, decimal amount)
    {
        if (amount < 0)
            throw new ArgumentException();
    }
}

public JournalCreditLine : IJournalEntryLine
{
    public JournalCreditLine(Account account, decimal amount)
    {
        if (amount > 0)
            throw new ArgumentException();
    }
}

public JournalEntry(... ICollection<JournalDebitLine> debits, ICollection<JournalCreditLine> credits)
{
    //no need to validate debits and credits, the responsibility is taken by the line types. also JournalEntry does not need to check each line, the types already ensure that.
}

public Report CreateExcelReport(ICollection<IJournalEntryLine> entryLines)
{
    foreach(var line in entryLines)
        Write(line.Amount); //
}```

File: source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs [hunk 2]
```diff
@@ -39,5 +43,28 @@ public JournalEntry(DateTime date, [NotNull] List<JournalEntryLine> debits, [Not
...
+
+        private static void EnsureParamNotNull(object param, string paramName, string exceptionMessage)
```
Comment (itiannerona) [inline source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs]: I think ``EnsureArgsNotNull`` will be accurate.

File: source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs [hunk 3]
```diff
@@ -39,5 +43,28 @@ public JournalEntry(DateTime date, [NotNull] List<JournalEntryLine> debits, [Not
...
+        {
+            if (param is null)
+                throw new ArgumentNullException(paramName, exceptionMessage);
+        }
+
+        private static void EnsureParamNotEmpty<T>(IEnumerable<T> collection, string paramName, string exceptionMessage)
```
Comment (itiannerona) [inline source/Pacioli/Pacioli.Lib/Models/JournalEntry.cs]: I think ``EnsureArgsNotEmpty`` will be accurate.

**Review by itiannerona — COMMENTED** [review]
Minor naming stuff and one refactor that we agreed on.

File: source/Pacioli/Pacioli.Lib/Models/Journal.cs [hunk 4]
```diff
@@ -1,13 +0,0 @@
-﻿using System.Linq;
-
-namespace Pacioli.Lib.Models
-{
-    public class Journal
```
Comment (maratmuginov) [inline source/Pacioli/Pacioli.Lib/Models/Journal.cs:5]: Journal was deleted as it was no longer covered by tests. JournalEntry now checks for balance.

**Review by itiannerona — APPROVED** [review]

**Pull request merged**